python/ray/data/datasource/file_based_datasource.py

```diff
@@ -225,24 +211,22 @@ def _resolve_paths_and_filesystem(
 
     resolved_paths = []
     for path in paths:
-        is_url = False
-        if filesystem is not None:
-            # If we provide a filesystem, _resolve_filesystem_and_path will not
-            # slice off the protocol from the provided URI/path when resolved.
-            is_url = _is_url(path)
-            path = _unwrap_protocol(path)
-        resolved_filesystem, resolved_path = _resolve_filesystem_and_path(
-            path, filesystem)
+        try:
+            resolved_filesystem, resolved_path = _resolve_filesystem_and_path(
+                path, filesystem)
+        except pa.lib.ArrowInvalid as e:
+            if "Cannot parse URI" in str(e):
```

Comment:
This seems pretty sketchy. Is there a way we can just always encode the url? If not, is there a way we can avoid checking the string in the exception?

Reply:
>Is there a way we can just always encode the url?

That's what I initially did, but there were [concerns](https://github.com/ray-project/ray/pull/18440#discussion_r704729649) about some pyarrow filesystems doing their own URL-encoding and therefore double-encoding our paths. We could always encode before and decode after, but some filesystems (such as local filesystems or fsspec filesystems) may not like pre-encoded paths.

>If not, is there a way we can avoid checking the string in the exception?

The most specific error type indicating a parsing error is `pa.lib.ArrowInvalid`, and pyarrow appears to rely on this kind of error message substring check internally for certain error types, so no other clearly better route immediately comes to mind. We could copy all of the `_resolve_filesystem_and_path` code and add this exception check directly to the `from_uri()` parsing (as I did earlier in this PR), but we'll still be matching on the error string, so it didn't seem worth it to copy the entire helper function over for logic that could easily wrap the helper. I'm going to open up an Arrow ticket for this, so (IMO) we should consider this a temporary workaround of an Arrow bug and not invest much time in creating a long-term workaround.